### PR TITLE
Fixes for reflectometry workflow algorithms history

### LINKED ISSUE #27633: Transmission WS has No History
**Original reporter:** [ISIS]/[Jos Cooper] on the [Mantid Forum](https://forum.mantidproject.org/t/createtransmissionworkspace-doesnt-show-history/620)

### Expected behavior
Any Workspace should have a history associated with its creation.
### Actual behavior
Transmission Workspaces have no history
### Steps to reproduce the behavior
Turn on the Archive in Manage User Directories.
In the Reflectometry GUI, in the Process Runs Table enter Run 57576, 2.3 degrees , 1st and 2nd Transmisson runs 57573 and 57574  (Instrument INTER).
Click Process and in thw Workspaces Toolbox in the main Mantid Window right-click on one of the TRANS workspaces and select "Show History".
You may get an error message: "Could not open history of 'TRANS_LAM_13463_13464'. 
ValueError: No history found on the given workspace"
### Platforms affected
4.2 Plot and WB

### PR DESCRIPTION
**Pull request by gemmaguest**
**Report to:** Jos

This PR fixes some issues with workspace history for interim output workspaces from the ISIS Reflectometry workflow algorithms. It includes:
- Addition of output properties for the interim workspaces to replace direct use of the ADS.
- Improved grouping of TOF workspaces. Previously only the input TOF workspaces were grouped, but now monitors are also grouped.
- A bug fix where the transmission workspace was being dumped before it was rebinned in `ReflectometryReductionOne::transmissionCorrection`, so there were minor binning differences in the workspace seen in the ADS to the one that was used in the correction.
- Minor renaming and tidying.

**To test:**
Set the default instrument to `INTER`. For each case in the table below:
- Run the code in the table to set up the arguments, and then use them to run the algorithm like this:
  ```
  ReflectometryISISLoadAndProcess(**args)
  ```
- Check the history of `IvsQ_binned_13460`. The following mandatory outputs should be set for every case:
  - OutputWorkspaceBinned=`IvsQ_binned_13460`
  - OutputWorkspace=`IvsQ_13460`
- The additional optional outputs that should also be set for each case are listed in the table. You can use this script will print the output properties:
  ```
  properties=['OutputWorkspaceBinned', 'OutputWorkspace', 'OutputWorkspaceWavelength',, 'OutputWorkspaceTransmission', 'OutputWorkspaceFirstTransmission', 'OutputWorkspaceSecondTransmission']
  for property in properties:
      print(property + ' = ' + mtd['IvsQ_binned_13460'].getHistory().getAlgorithmHistory(0).getPropertyValue(property))
  ```
- Where a `TRANS_LAM...` workspace is output, check that it has history. Note that the loaded runs in TOF do not have history.

Script | Optional outputs
-------|-----------------
`args={'Debug':0, 'InputRunList':'13460'}` | None
`args={'Debug':1, 'InputRunList':'13460'}` | OutputWorkspaceWavelength=`IvsLam_13460`
`args={'Debug':0, 'InputRunList':'13460', 'FirstTransmissionRunList':'13463'}` | OutputWorkspaceTransmission=`TRANS_LAM_13463`
`args={'Debug':1, 'InputRunList':'13460', 'FirstTransmissionRunList':'13463'}` | OutputWorkspaceWavelength=`IvsLam_13460` OutputWorkspaceTransmission=`TRANS_LAM_13463`
`args={'Debug':0, 'InputRunList':'13460', 'FirstTransmissionRunList':'13463', 'SecondTransmissionRunList':'13464'}` | OutputWorkspaceTransmission=`TRANS_LAM_13463_13464`
`args={'Debug':1, 'InputRunList':'13460', 'FirstTransmissionRunList':'13463', 'SecondTransmissionRunList':'13464'}`| OutputWorkspaceWavelength=`IvsLam_13460` OutputWorkspaceTransmission=`TRANS_LAM_13463_13464` OutputWorkspaceFirstTransmission=`TRANS_LAM_13463` OutputWorkspaceSecondTransmission=`TRANS_LAM_13464`

Fixes #27633 

Release notes to follow

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by giovannidisiena — CHANGES_REQUESTED** [review]
This looks good to me so far but there are a lot of changes here so might be worth having a second reviewer.

File: qt/widgets/common/inc/MantidQtWidgets/Common/DataProcessorUI/ProcessingAlgorithm.h [hunk 1]
```diff
@@ -28,11 +28,12 @@ class EXPORT_OPT_MANTIDQT_COMMON ProcessingAlgorithm
...
   ProcessingAlgorithm(QString name, std::vector<QString> prefix,
                       std::size_t postprocessedOutputPrefixIndex,
-                      std::set<QString> blacklist = std::set<QString>());
+                      std::set<QString> blacklist = std::set<QString>(),
+                      const int version = -1);
   // Delegating constructor
   ProcessingAlgorithm(QString name, QString const &prefix,
                       std::size_t postprocessedOutputPrefixIndex,
-                      QString const &blacklist = "");
+                      QString const &blacklist = "", const int version = -1);
```
[the remark this answers is not in the document]
Comment (gemmaguest) [reply qt/widgets/common/inc/MantidQtWidgets/Common/DataProcessorUI/ProcessingAlgorithm.h]: Normally I'd agree, but the `-1` is the convention used in the algorithms so this just passes it directly through.

**Comment by gemmaguest** [comment]
I'm just going to rebase on master and update the release notes

**Comment by gemmaguest** [comment]
> This looks good to me so far but there are a lot of changes here so might be worth having a second reviewer.

This doesn't touch any core components so I think the normal gatekeeper review should be fine. Note that the data processor code that has been changed is not currently used anyway! The PR does have the potential to break reflectometry scripts but scientists are aware of that and I will be pushing heavily for them to beta test it.

**Comment by gemmaguest** [comment]
I'll take a look at the conflicts

**Comment by giovannidisiena** [comment]
Having some problems with my archive so going to clear my assignment to allow someone else to do functional testing.

**Comment by gemmaguest** [comment]
> Having some problems with my archive so going to clear my assignment to allow someone else to do functional testing.

@giovannidisiena The files in the ISIS sample data should be sufficient so you shouldn't need the archive?

**Review by giovannidisiena — APPROVED** [review]
All appears good to me :shipit:

**Comment by NickDraper** [comment]
Final build failure was just network connectivity with the build slave

**Pull request merged**